src/App.tsx

```diff
@@ -1,24 +1,278 @@
+  const [errorType, setErrorType] = useState('');
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [hasError, setHasError] = useState(false);
+  const [isTodoAdding, setIsTodoAdding] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [updatingTodoIds, setUpdatingTodoIds] = useState<number[]>([]);
+  const [removingTodoIds, setRemovingTodoIds] = useState<number[]>([]);
+  const [selectedFilter, setSelectedFilter] = useState(FilterValues.ALL);
+  const activeTodos = [...todos].filter(todoItem => !todoItem.completed);
+  const isEveryCompleted = todos.every(todoItem => todoItem.completed);
+  const hasCompletedTodo = todos.some(todo => todo.completed);
+
+  const getTodosFromServer = async () => {
+    try {
+      const response = await getTodos(USER_ID);
+
+      setTodos(response);
+    } catch (error) {
+      setHasError(true);
+      setErrorType('upload');
+    }
+  };
+
+  const handleAddTodo = useCallback(async () => {
+    const newTodoToFetch = {
+      userId: USER_ID,
+      title,
+      completed: false,
+    };
+
+    const newTodoToShow = {
+      ...newTodoToFetch,
+      id: 0,
+    };
+
+    setIsTodoAdding(true);
+    setTempTodo(newTodoToShow);
+
+    try {
+      const addedTodo = await addTodo(newTodoToFetch);
+
+      setTodos(prevTodos => [...prevTodos, addedTodo]);
+
+      setHasError(false);
+    } catch (error) {
+      setHasError(true);
+      setErrorType('add');
+    } finally {
+      setIsTodoAdding(false);
+      setTempTodo(null);
+    }
+  }, [title]);
+
+  const handleDeleteTodo = useCallback(async (todoId: number) => {
+    setRemovingTodoIds(prevTodoIds => [...prevTodoIds, todoId]);
+
+    try {
+      await removeTodo(todoId);
+
+      const updatedTodos = [...todos].filter(todo => todo.id !== todoId);
+
+      setTodos(updatedTodos);
+    } catch (error) {
+      setHasError(true);
+      setErrorType('delete');
+    } finally {
+      setRemovingTodoIds(prevTodoIds => prevTodoIds
+        .filter((id) => id !== todoId));
+    }
+  }, [todos]);
+
+  const clearCompletedTodos = useCallback(async () => {
+    const completedTodosIds = [...todos]
+      .filter(todo => todo.completed)
+      .map((todo) => todo.id);
+
+    setRemovingTodoIds(completedTodosIds);
+
+    Promise.all(completedTodosIds.map((id) => removeTodo(id)))
+      .then(() => {
+        setTodos(activeTodos);
+      })
+      .catch(() => {
+        setHasError(true);
+        setErrorType('delete');
+      })
+      .finally(() => setRemovingTodoIds(prevTodoIds => prevTodoIds
+        .filter((id) => !completedTodosIds.includes(id))));
+  }, [hasCompletedTodo, activeTodos]);
+
+  const handleChangeStatus = useCallback(async (todo: Todo) => {
+    setUpdatingTodoIds(prevTodoIds => [...prevTodoIds, todo.id]);
+
+    const updatedTodo = {
+      ...todo,
+      completed: !todo.completed,
+    };
+
+    try {
+      await updateTodo(todo.id, updatedTodo);
```

Comment:
![image](https://user-images.githubusercontent.com/122989585/224400406-99631915-d807-4515-92a0-2e6290e6332d.png)
```suggestion
      const updatedTodo = await updateTodo(todo.id, { userId: id.., completed: true/false });
```